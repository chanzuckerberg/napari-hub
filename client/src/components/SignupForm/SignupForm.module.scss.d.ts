export type Styles = {
  red: string;
  signupForm: string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
