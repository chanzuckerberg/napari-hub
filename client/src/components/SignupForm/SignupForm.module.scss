.signupForm {
  @apply w-full;
}

.red {
  @apply text-red-500;
}
