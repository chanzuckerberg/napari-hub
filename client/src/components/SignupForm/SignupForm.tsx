import clsx from 'clsx';
import React, { useRef, useState } from 'react';

import { ColumnLayout } from '@/components/common';

// MailChimp form target
const MC_URL = new URL('https://dev.us5.list-manage.com/subscribe/post');
MC_URL.search = new URLSearchParams({
  u: '79f50c0d0474ff43dffedabda',
  id: '63bf16756e',
}).toString();

interface Props {
  onSubmit?: (event: React.FormEvent) => void;
}

/**
 * Signup form to capture email address and send it to MailChimp
 */
export function SignupForm({ onSubmit }: Props) {
  const [email, setEmail] = useState('');
  const [error, setError] = useState('');
  const emailRef = useRef<HTMLInputElement | null>(null);

  const validate = () => {
    const validityState = emailRef.current?.validity;

    if (validityState?.valueMissing) {
      setError('Please enter an email address');
      return false;
    }
    if (validityState?.typeMismatch) {
      setError('Please provide a valid email address');
      return false;
    }

    setError(''); // no error
    return true;
  };

  const handleSubmit = (event: React.FormEvent) => {
    const isValid = validate();

    if (!isValid) {
      event.preventDefault(); // do not submit form
    } else if (onSubmit) {
      onSubmit(event);
    }
  };

  return (
    <ColumnLayout className="bg-napari-light p-6 xl:px-0 xl:py-12">
      <div className="col-span-2 md:col-span-3 3xl:col-start-2">
        <h3 className="prose-lg font-semibold mb-4">
          Sign up to receive updates
        </h3>
        <form
          onSubmit={handleSubmit}
          className="grid grid-cols-2 xl:grid-cols-napari-3 gap-y-0.5 md:gap-x-12"
          noValidate
          action={MC_URL.href}
          method="post"
          target="_blank"
        >
          <input
            ref={emailRef}
            type="email"
            name="EMAIL"
            aria-label="email address"
            placeholder="email"
            value={email}
            onChange={(event) => setEmail(event.target.value)}
            required
            className={clsx(
              // font & colors
              'prose-sm bg-transparent focus:outline-none',
              // sizing
              'h-10 xl:h-[35px] col-span-2 md:col-span-1 xl:col-span-2',
              // underline
              'border-b',
              !error ? 'border-black' : 'border-napari-error',
            )}
            data-testid="emailField"
          />
          <span
            data-testid="emailError"
            className="text-xs align-text-top h-[1.5em] text-napari-error md:row-start-2 col-span-2 md:col-span-1 xl:col-span-2"
          >
            {error}
          </span>
          <input
            type="submit"
            name="subscribe"
            value="Subscribe"
            className={clsx(
              // font & colors
              'prose-sm font-semibold bg-napari-primary',
              // sizing
              'h-[35px] col-span-2 md:col-span-1',
              // border & interaction
              'border-none cursor-pointer',
            )}
            data-testid="submitButton"
          />
        </form>
      </div>
    </ColumnLayout>
  );
}
