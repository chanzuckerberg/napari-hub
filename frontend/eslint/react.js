module.exports = {
  extends: ['airbnb/hooks'],

  plugins: ['napari-hub'],

  rules: {
    'napari-hub/no-root-mui-import': 'error',

    /*
      Prop types aren't necessary since we have TypeScript interfaces for
      prop types.
    */
    'react/prop-types': 'off',

    /*
      React automatically adds the import as part of the new JSX transform:
      https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html#nextjs
     */
    'react/react-in-jsx-scope': 'off',

    /*
      Prop spreading is very useful as a simple way for passing multiple
      props to a component. However, we should still be careful about passing
      unnecessary props.
    */
    'react/jsx-props-no-spreading': 'off',

    /*
      This rule isn't really needed anymore since we can use ES6 object defaults
      for props. There are also plans to deprecate the use of `defaultProps`:
      https://github.com/reactjs/rfcs/pull/107
    */
    'react/require-default-props': 'off',
  },
};
