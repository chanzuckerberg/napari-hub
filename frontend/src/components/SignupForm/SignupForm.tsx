import Button from '@material-ui/core/Button';
import { FormHelperTextProps } from '@material-ui/core/FormHelperText';
import TextField from '@material-ui/core/TextField';
import clsx from 'clsx';
import { useTranslation } from 'next-i18next';
import React, { useRef, useState } from 'react';

import { ColumnLayout } from '@/components/common/ColumnLayout';
import { usePlausible } from '@/hooks';

// MailChimp form target
const MC_URL = new URL('https://dev.us5.list-manage.com/subscribe/post');
MC_URL.search = new URLSearchParams({
  u: '79f50c0d0474ff43dffedabda',
  id: '63bf16756e',
}).toString();

interface Props {
  onSubmit?: (event: React.FormEvent) => void;
  variant?: 'default' | 'home';
}

/**
 * Signup form to capture email address and send it to MailChimp
 */
export function SignupForm({ onSubmit, variant = 'default' }: Props) {
  const [t] = useTranslation(['footer']);
  const [email, setEmail] = useState('');
  const [error, setError] = useState('');
  const emailRef = useRef<HTMLInputElement | null>(null);
  const plausible = usePlausible();
  const isHome = variant === 'home';

  const validate = () => {
    const validityState = emailRef.current?.validity;

    if (validityState?.valueMissing) {
      setError(t('footer:signUp.missingEmail'));
      return false;
    }
    if (validityState?.typeMismatch) {
      setError(t('footer:signUp.invalidEmail'));
      return false;
    }

    setError(''); // no error
    return true;
  };

  const handleSubmit = (event: React.FormEvent) => {
    const isValid = validate();

    if (isValid) {
      onSubmit?.(event);
      plausible('Signup');
    } else {
      event.preventDefault(); // do not submit form
    }
  };

  return (
    <ColumnLayout
      className="bg-napari-light p-6 screen-495:p-12"
      // Use 3-column layout instead of 4-column layout.
      classes={isHome ? { fourColumn: 'screen-1150:grid-cols-napari-3' } : {}}
    >
      <div className="col-span-2 screen-495:col-span-3 screen-1425:col-start-2">
        <h3 className="text-lg font-semibold mb-1 screen-495:mb-4">
          {t('footer:signUp.title')}
        </h3>
        <form
          onSubmit={handleSubmit}
          noValidate
          action={MC_URL.href}
          method="post"
          target="_blank"
          className={clsx(
            // grid
            'grid grid-cols-2 screen-875:grid-cols-napari-3',
            // spacing
            'screen-495:gap-x-12',
            // make room for error message in vertical layout
            error ? 'gap-y-8' : 'gap-y-3.5',
          )}
        >
          <TextField
            inputRef={emailRef}
            error={error !== ''}
            helperText={error}
            type="email"
            name="EMAIL"
            aria-label={t('footer:ariaLabels.emailAddress')}
            placeholder="email"
            value={email}
            onChange={(event) => setEmail(event.target.value)}
            required
            margin="none"
            className={clsx(
              // font
              'text-sm',
              // make underline align with submit button
              'pt-0.5',
              // sizing
              'h-8 screen-495:h-[35px] col-span-2 screen-495:col-span-1 screen-875:col-span-2',
            )}
            inputProps={{ 'data-testid': 'emailField' }}
            FormHelperTextProps={
              { 'data-testid': 'emailError' } as FormHelperTextProps
            }
          />
          <Button
            type="submit"
            color="primary"
            name="subscribe"
            variant="contained"
            disableElevation
            className={clsx(
              // font & colors
              'text-sm font-semibold',
              // sizing
              'h-[35px] col-span-2 screen-495:col-span-1',
            )}
            data-testid="submitButton"
          >
            {t('footer:signUp.subscribe')}
          </Button>
        </form>
      </div>
    </ColumnLayout>
  );
}
