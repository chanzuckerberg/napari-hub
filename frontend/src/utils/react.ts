import { ReactElement, ReactNode } from 'react';

/**
 * Checks if a React node is an element. An element is an instantiated
 * component.
 *
 * @param node The React node
 * @returns True if node is an element
 */
export function isReactElement(node: ReactNode): node is ReactElement {
  const element = node as ReactElement | null | undefined;
  return !!(element?.type && element?.props);
}
