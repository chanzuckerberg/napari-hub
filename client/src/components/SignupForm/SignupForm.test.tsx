import { render, screen } from '@testing-library/react';

import { SignupForm } from './SignupForm';

describe('<SignupForm />', () => {
  it('should match snapshot', () => {
    const component = render(<SignupForm />);
    expect(component).toMatchSnapshot();
  });
});
