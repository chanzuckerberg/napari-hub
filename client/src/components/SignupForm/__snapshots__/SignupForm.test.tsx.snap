// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignupForm /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="bg-napari-light p-6 xl:px-0 xl:py-12 grid justify-center gap-6 md:gap-12 grid-cols-2 screen-875:grid-cols-napari-3 screen-1150:grid-cols-napari-4 screen-1425:grid-cols-napari-5"
  >
    <div
      class="col-span-2 md:col-span-3 3xl:col-start-2"
    >
      <h3
        class="prose-lg font-semibold mb-4"
      >
        Sign up to receive updates
      </h3>
      <form
        action="https://dev.us5.list-manage.com/subscribe/post?u=79f50c0d0474ff43dffedabda&id=63bf16756e"
        class="grid grid-cols-2 xl:grid-cols-napari-3 gap-y-0.5 md:gap-x-12"
        method="post"
        novalidate=""
        target="_blank"
      >
        <input
          aria-label="email address"
          class="prose-sm bg-transparent focus:outline-none h-10 xl:h-[35px] col-span-2 md:col-span-1 xl:col-span-2 border-b border-black"
          data-testid="emailField"
          name="EMAIL"
          placeholder="email"
          required=""
          type="email"
          value=""
        />
        <span
          class="text-xs align-text-top h-[1.5em] text-napari-error md:row-start-2 col-span-2 md:col-span-1 xl:col-span-2"
          data-testid="emailError"
        />
        <input
          class="prose-sm font-semibold bg-napari-primary h-[35px] col-span-2 md:col-span-1 border-none cursor-pointer"
          data-testid="submitButton"
          name="subscribe"
          type="submit"
          value="Subscribe"
        />
      </form>
    </div>
  </div>
</DocumentFragment>
`;
